src/dmd/sideeffect.d

```diff
@@ -356,6 +353,28 @@ bool discardValue(Expression e)
         if (!hasSideEffect(e))
             break;
         return false;
+    case EXP.identity, EXP.notIdentity:
+    case EXP.equal, EXP.notEqual:
+        /*
+            `[side effect] == 0`
+            Technically has a side effect but is clearly wrong;
+        */
+        BinExp tmp = e.isBinExp();
+        assert(tmp);
+
+        e.error("`%s` may not be discarded, since it is likely a mistake", e.toChars());
```

Comment:
Language is a bit coarse, and it's still not obvious what you are talking about based on the testsuite diff alone.

Reply:
Examples from other language compilers.
```
`fun().i == 4` value computed is not used

`fun().i == 4` equality comparison result unused
```